pyteal/ast/router.py

```diff
@@ -0,0 +1,541 @@
+from pyteal.ast.methodsig import MethodSignature
+from pyteal.ast.naryexpr import And, Or
+from pyteal.ast.txn import Txn
+from pyteal.ast.return_ import Approve
+
+
+class CallConfig(Enum):
+    """
+    CallConfigs: a "bitset"-like class for more fine-grained control over
+    `call or create` for a method about an OnComplete case.
+
+    This enumeration class allows for specifying one of the four following cases:
+    - CALL
+    - CREATE
+    - ALL
+    - NEVER
+    for a method call on one on_complete case.
+    """
+
+    NEVER = 0
+    CALL = 1
+    CREATE = 2
+    ALL = 3
+
+    def __or__(self, other: object) -> "CallConfig":
+        if not isinstance(other, CallConfig):
+            raise TealInputError("CallConfig must be compared with same class")
+        return CallConfig(self.value | other.value)
+
+    def __and__(self, other: object) -> "CallConfig":
+        if not isinstance(other, CallConfig):
+            raise TealInputError("CallConfig must be compared with same class")
+        return CallConfig(self.value & other.value)
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, CallConfig):
+            raise TealInputError("CallConfig must be compared with same class")
+        return self.value == other.value
+
+
+CallConfig.__module__ = "pyteal"
+
+
+@dataclass(frozen=True)
+class CallConfigs:
+    """
+    CallConfigs keep track of one method registration's CallConfigs for all OnComplete cases.
+    """
+
+    no_op: CallConfig = field(kw_only=True, default=CallConfig.CALL)
+    opt_in: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    close_out: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    clear_state: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    update_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+    delete_application: CallConfig = field(kw_only=True, default=CallConfig.NEVER)
+
+    def is_never(self) -> bool:
+        return all(map(lambda cc: cc == CallConfig.NEVER, astuple(self)))
+
+    def _oc_under_call_config(self, call_config: CallConfig) -> list[EnumInt]:
+        if not isinstance(call_config, CallConfig):
+            raise TealInputError(
+                "generate condition based on CallConfigs should be based on CallConfig"
+            )
+        config_oc_pairs: list[tuple[CallConfig, EnumInt]] = [
+            (self.no_op, OnComplete.NoOp),
+            (self.opt_in, OnComplete.OptIn),
+            (self.close_out, OnComplete.CloseOut),
+            (self.clear_state, OnComplete.ClearState),
+            (self.update_application, OnComplete.UpdateApplication),
+            (self.delete_application, OnComplete.DeleteApplication),
+        ]
+        return [
+            oc
+            for oc_config, oc in config_oc_pairs
+            if (oc_config & call_config) != CallConfig.NEVER
+        ]
+
+    def partition_oc_by_clear_state(self, cc: CallConfig) -> tuple[bool, list[EnumInt]]:
```

Comment:
This is no longer a partition, so maybe rename to something like `_has_non_clears_n_clear_states` (admittedly, very ugly name)

Reply:
I separated them into two methods, and I think it might lead to better readability.